### ISSUE TITLE
A handful of questions regarding exchangelib's goals

### ISSUE THREAD

**Issue by SamCB** [ecederstrand/exchangelib#12]
Hi there,

I see that there's continued development of exchangelib. I'm wondering if there is a plan for what this library is eventually going to be? 

Is the aim to completely cover all the operations of EWS? Is there any plan to follow the EWS Managed API? Is there any expectations for contributions if you're accepting them?

I might be spending a bit of time writing some stuff for this library (have a bit of experimentation to do with pull subscriptions in my fork) which I might be able to contribute back. Of course it'll be a lot better if I can write it in a way that fits in with the overall vision.

**Comment by ecederstrand** [comment]
First, thanks for your interest!

At bit of history. This library started out as a PHP script around 2008 before anyone had made a PHP library for EWS. I published [this article](https://www.howtoforge.com/talking-soap-with-exchange) describing how to start, as I wanted to share my knowledge.

In the meantime, I got fed up with PHP and learned Python. I still needed an EWS library and ported my PHP code. I needed something that was fast and easy to use for calendar access which is my primary need. My company synchronizes millions of calendar items to 10.000's of accounts on 100's of servers with minimal delay. Earlier this year, looking around PyPI and answering questions on StackOverflow, I saw that there was no easy-to-use EWS library for Python and I decided to clean up my work and open-source it.

It's not difficult to create a library that consumes the WSDL and allows full access to everything. But fully auto-generating code easily becomes magic for the uninitiated, quirks are difficult to handle and consumer code becomes awkward and a mess. WSDL files are not always accessible or point to a wrong version. The library needs to "Just Work", just like Outlook "Just Works". Hence, there's a lot of tasting, guessing and workarounds for quirks and common server misconfigurations behind the scenes.

I want to provide something that is easy to use, as fast as possible and intuitive for a Python programmer. It solves the common cases, possibly sacrificing exotic functionality. I don't mind extending support beyond my own needs (I added inbox, tasks and contacts support because it was easy), but I'm also aware that a library that does everything quickly becomes a library that does nothing. Simple requirements should mean simple consumer code.

So, to answer your questions;

No, I'm not aiming to cover the full EWS functionality. If it can be done without sacrificing code readability and library usability, I'm not against it.

No, I'm not aiming to follow the Managed API. I haven't used it myself and the examples I have seen have been over-complicated for even simple needs.

I'm very welcome to contributions. New code should leave code quality in a better shape than before - faster, more "pythonic", better documented, better tested. I'm learning along the way, too.
